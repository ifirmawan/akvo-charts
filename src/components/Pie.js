import React from 'react';
import { useECharts } from '../hooks';
import styles from '../styles.module.css';

const getOptions = ({ dimensions = [] }) => {
  const itemName = dimensions[0];
  const value = dimensions.slice(1);
  return {
    series: [
      {
        type: 'pie',
        encode: {
          itemName,
          value
        }
      }
    ]
  };
};

const Pie = ({ config, data }) => {
  const chartRef = useECharts({
    config,
    data,
    getOptions: ({ dimensions }) => getOptions({ dimensions })
  });

  return (
    <div
      ref={chartRef}
      role="figure"
      className={styles.container}
    />
  );
};

export default Pie;
