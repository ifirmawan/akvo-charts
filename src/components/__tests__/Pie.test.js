import React from 'react';
import { render, screen } from '@testing-library/react';
import Pie from '../Pie';
import renderer from 'react-test-renderer';

describe('Pie chart', () => {
  test('renders Pie component with 2d array data format', () => {
    const data = [
      ['product', '2015', '2016', '2017'],
      ['Matcha Latte', 43.3, 85.8, 93.7],
      ['Milk Tea', 83.1, 73.4, 55.1],
      ['Cheese Cocoa', 86.4, 65.2, 82.5],
      ['Walnut Brownie', 72.4, 53.9, 39.1]
    ];

    const config = {
      title: 'Pie Chart Example'
    };

    render(
      <Pie
        config={config}
        data={data}
      />
    );

    const chartContainer = screen.getByRole('figure');
    expect(chartContainer).toBeInTheDocument();
  });

  test('renders Pie component with row based key-value format (object array)', () => {
    const data = [
      { product: 'Matcha Latte', count: 823, score: 95.8 },
      { product: 'Milk Tea', count: 235, score: 81.4 },
      { product: 'Cheese Cocoa', count: 1042, score: 91.2 },
      { product: 'Walnut Brownie', count: 988, score: 76.9 }
    ];

    const config = {
      title: 'Pie Chart Example'
    };

    render(
      <Pie
        config={config}
        data={data}
      />
    );

    const chartContainer = screen.getByRole('figure');
    expect(chartContainer).toBeInTheDocument();
  });

  test('renders Pie component with column based key-value format', () => {
    const data = {
      product: ['Matcha Latte', 'Milk Tea', 'Cheese Cocoa', 'Walnut Brownie'],
      count: [823, 235, 1042, 988],
      score: [95.8, 81.4, 91.2, 76.9]
    };

    const config = {
      title: 'Pie Chart Example'
    };

    render(
      <Pie
        config={config}
        data={data}
      />
    );

    const chartContainer = screen.getByRole('figure');
    expect(chartContainer).toBeInTheDocument();
  });

  test('matches Pie snapshot', () => {
    const data = [
      ['product', '2015', '2016', '2017'],
      ['Matcha Latte', 43.3, 85.8, 93.7],
      ['Milk Tea', 83.1, 73.4, 55.1],
      ['Cheese Cocoa', 86.4, 65.2, 82.5],
      ['Walnut Brownie', 72.4, 53.9, 39.1]
    ];

    const config = {
      title: 'Pie Chart Example'
    };

    const tree = renderer
      .create(
        <Pie
          config={config}
          data={data}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
