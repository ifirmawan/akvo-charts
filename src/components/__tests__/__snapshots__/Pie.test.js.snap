// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pie chart matches Pie snapshot 1`] = `
<div
  className="container"
  role="figure"
/>
`;
